geth/node/manager.go

```diff
@@ -518,3 +519,64 @@ func (m *NodeManager) isNodeAvailable() error {
 
 	return nil
 }
+
+// tickerResolution is the delta to check blockchain sync progress.
+const tickerResolution = time.Second
+
+// Sync waits until blockchain synchronization and returns.
+// Timeout must be at least two seconds and zero timeout
+// means waiting infinitely.
+func (m *NodeManager) Sync(timeout time.Duration) error {
+	// We need to have a larger timeout than ticker delta here
+	// unless we use zero which means infinite timeout.
+	if timeout < tickerResolution*2 && timeout != 0 {
+		return errors.New("Sync timeout can only be zero (infinite) or at least two seconds")
+	}
+	// Don't wait for any blockchain sync for the
+	// local private chain as blocks are never mined.
+	if m.config.NetworkID == params.StatusChainNetworkID {
+		return nil
+	}
+	if m.lesService == nil {
+		return errors.New("LightEthereumService is nil")
+	}
+	return m.sync(timeout)
+}
+
+func (m *NodeManager) sync(timeout time.Duration) error {
+	if timeout == 0 {
+		// Wait for a year (infinite).
+		timeout = time.Hour * 8765
+	}
+
+	ticker := time.NewTicker(tickerResolution)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-time.After(timeout):
```

Comment:
I am sorry, I remember I suggested changing it to this form but in this case, it may be better to do it differently:
```go
var timeoutCh <-chan time.Time

if timeout == 0 {
  timeoutCh = make(chan time.Time)
} else {
  timeoutCh = time.After(timeout)
}

// below
select {
case <-timeoutCh:
  // ...
case <-ticker.C:
  // ...
```

And in this way you can avoid a hackish `one year timeout` :)

Reply:
Thank you, I was actually looking for something like this before but couldn't imagine :smile: